File: src/my_py_pkg/my_py_pkg/hardware_status_publisher.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from my_robot_interfaces.msg import HardwareStatus


class HardwareStatusPublisherNode(Node):
    def __init__(self):
        super().__init__("hardware_status_publisher")
        self.publisher_ = self.create_publisher(HardwareStatus, "hardware_status", 10)
        self.timer_ = self.create_timer(1.0, self.publish_hardware_status_callback)

    def publish_hardware_status_callback(self):
        msg = HardwareStatus()
        msg.temperature = 25.0  # Example temperature value
        msg.are_motors_ready = True  # Example motor readiness status
        msg.debug_message = "All systems operational."  # Example debug message
        self.publisher_.publish(msg)

def main(args=None):
    rclpy.init(args=args)
    node = HardwareStatusPublisherNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/my_cpp_pkg/src/hardware_status_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "my_robot_interfaces/msg/hardware_status.hpp"

class HardwareStatusPublisherNode : public rclcpp::Node
{
public:
    HardwareStatusPublisherNode() : Node("hardware_status_publisher")
    {
        publisher_ = this->create_publisher<my_robot_interfaces::msg::HardwareStatus>("hardware_status", 10);
        timer_ = this->create_wall_timer(
            std::chrono::seconds(1),
            std::bind(&HardwareStatusPublisherNode::publish_hardware_status, this));
        RCLCPP_INFO(this->get_logger(), "Hardware Status Publisher node has been started and is publishing hardware status.");
    }

private:
    void publish_hardware_status()
    {
        auto msg = my_robot_interfaces::msg::HardwareStatus();
        // Populate the message with hardware status information
        msg.temperature = 42.0; // Example temperature value
        msg.are_motors_ready = true; // Example motor readiness status
        msg.debug_message = "All systems operational"; // Example debug message
        publisher_->publish(msg);
    }

    rclcpp::Publisher<my_robot_interfaces::msg::HardwareStatus>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<HardwareStatusPublisherNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

